applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -895,276 +469,343 @@ static void cmd_send(struct k_work *work)
 	/** Format as TS 27.007 command V1 with verbose response format,
 	 *  based on current return of API nrf_modem_at_cmd() and MFWv1.3.x
 	 */
-	if (strlen(at_buf) > strlen(CRLF_STR)) {
+	if (strlen(buf) > strlen(CRLF_STR)) {
 		format_final_result(at_buf, strlen(at_buf), sizeof(at_buf));
-		(void)uart_send(at_buf, strlen(at_buf));
+		err = slm_uart_tx_write(buf, strlen(buf));
+		if (err) {
+			LOG_ERR("AT command response failed: %d", err);
+		}
 	}
 
 done:
-	(void)uart_receive_enable();
+	return;
 }
 
-static int cmd_rx_handler(uint8_t character)
+static size_t cmd_rx_handler(const uint8_t *buf, const size_t len)
 {
+	size_t processed;
 	static bool inside_quotes;
 	static size_t at_cmd_len;
-
-	/* Handle control characters */
-	switch (character) {
-	case 0x08: /* Backspace. */
-		/* Fall through. */
-	case 0x7F: /* DEL character */
-		if (at_cmd_len > 0) {
-			at_cmd_len--;
+	static uint8_t prev_character;
+	bool send = false;
+
+	for (processed = 0; processed < len && send == false; processed++) {
+
+		/* Handle control characters */
+		switch (buf[processed]) {
+		case 0x08: /* Backspace. */
+			/* Fall through. */
+		case 0x7F: /* DEL character */
+			if (at_cmd_len > 0) {
+				at_cmd_len--;
+			}
+			continue;
 		}
-		return 0;
-	}
 
-	/* Handle termination characters, if outside quotes. */
-	if (!inside_quotes) {
-		switch (character) {
-		case '\r':
+		/* Handle termination characters, if outside quotes. */
+		if (!inside_quotes) {
+			switch (buf[processed]) {
+			case '\r':
 #if defined(CONFIG_SLM_CR_TERMINATION)
-			goto send;
+				send = true;
 #else
-			break;
+				break;
 #endif
-		case '\n':
+			case '\n':
 #if defined(CONFIG_SLM_LF_TERMINATION)
-			goto send;
+				send = true;
 #elif defined(CONFIG_SLM_CR_LF_TERMINATION)
-			if (at_cmd_len > 0 && at_buf[at_cmd_len - 1] == '\r') {
-				at_cmd_len--; /* trim the CR char */
-				goto send;
-			}
+				if (at_cmd_len > 0 && prev_character == '\r') {
+					at_cmd_len--; /* trim the CR char */
+					send = true;
+				}
 #endif
```

Comment:
`case '\n':`  seem to be missing `break;` inside `#if defined`. It only sets a flag and flows into `case '\n'`

This #ifdef thing is hard to read. Please refactor to use `if (IS_ENABLED(...))`

```c
		if (!inside_quotes) {
			switch (buf[processed]) {
			case '\r':
				if (IS_ENABLED(CONFIG_SLM_CR_TERMINATION)) {
					send = true;
				}
				break;
			case '\n':
				if (IS_ENABLED(CONFIG_SLM_LF_TERMINATION)) {
					send = true;
				} else if (IS_ENABLED(CONFIG_SLM_CR_LF_TERMINATION)) {
					if (at_cmd_len > 0 && prev_character == '\r') {
						at_cmd_len--; /* trim the CR char */
						send = true;
					}
				}
				break;
			}
		}
```

Reply:
Refactored as suggested.